### PR TITLE
Cache more things on CircleCI

### PR DESCRIPTION
**Pull request by underyx**

### CODE REVIEW

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/d01e3bf5b29fa1f5114ded010947ef2061cb9b96

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/e012fc692f8d842119e2b86cc9ef13a33ff06dc2

**Comment by AbdealiJK** [comment]
@underyx If you're free enough, could you also add:
 - bundler cacheing
 - Adding an arg to pytest to fail if cov is not 100% https://github.com/coala-analyzer/coala/issues/2126

They're 2 minute additions and would be awesome to have

**Comment by underyx** [comment]
I was already working on bundler caching but something deep is wrong with it :/

I can definitely do https://github.com/coala-analyzer/coala/issues/2126

Will look at this PR again later today.

File: .ci/deps.sh [hunk 1]
```diff
@@ -105,3 +105,9 @@ fi
 # Tailor (Swift) commands
 curl -fsSL https://tailor.sh/install.sh | sed 's/read -r CONTINUE < \/dev\/tty/CONTINUE=y/' > install.sh
 sudo bash install.sh
+
+# making coala cache the dependencies downloaded upon first run
+echo '' > dummy
+coala-ci --bears CheckstyleBear --files dummy --no-config || true
```
Comment (AbdealiJK) [inline .ci/deps.sh]: Use `/dev/null` instead (Minor)

Also, why `|| true` ?

File: tests/java/CheckstyleBearTest.py [hunk 2]
```diff
@@ -31,3 +32,13 @@ def test_with_custom_configfile(self):
         self.section["checkstyle_configs"] = self.empty_config
         self.check_validity(self.uut, [], self.good_file)
         self.check_validity(self.uut, [], self.bad_file)
+
+
+def test_known_checkstyle(monkeypatch):
```
Comment (AbdealiJK) [inline tests/java/CheckstyleBearTest.py]: What is monkeypatch ?
Possibly use unittest.mock.patch ? Then it can be inside the TestCase class and that's what we do in other places and also you wouldn't need to change `from bears.java.CheckstyleBear import CheckstyleBear`

File: .ci/deps.sh (hunk 1 above)
Comment (underyx) [inline .ci/deps.sh]: @AbdealiJK I tried /dev/null first, but coala couldn't check it since it's
outside the working directory.

`|| true` is there to make sure the build doesn't get interrupted if a bear
returns an error for an empty file. (For instance YamlLintBear I think
would complain that the front matter is missing.) The purpose of the
command is to get the deps and that would be done regardless of exit code.

File: bears/js/JSONFormatBear.py [hunk 3]
```diff
@@ -10,10 +10,7 @@
 
 
 class JSONFormatBear(LocalBear):
-    try:
-        DecodeError = json.decoder.JSONDecodeError
-    except AttributeError:
-        DecodeError = ValueError
+    DecodeError = json.decoder.JSONDecodeError
```
Comment (AbdealiJK) [inline bears/js/JSONFormatBear.py]: It does not ahttps://docs.python.org/3/library/json.html#json.JSONDecodeError - JSONDecodeError is new in 3.5
Hence it wont exist in 3.4 and 3.3. Which is why this is there.

File: tests/java/CheckstyleBearTest.py (hunk 2 above)
Comment (underyx) [inline tests/java/CheckstyleBearTest.py]: monkeypatch is a fixture provided by pytest

I don't want this to be inside that TesCase since it's testing a different
unit. The import would still need to be changed since the old import gets
only the bear class, and this is testing a module-level function.

File: bears/js/JSONFormatBear.py (hunk 3 above)
Comment (underyx) [inline bears/js/JSONFormatBear.py]: Yep, I saw this in the AppVeyor build, but it was already 3:30am. How can I
test that `except` branch then?

File: tests/java/CheckstyleBearTest.py (hunk 2 above)
Comment (AbdealiJK) [inline tests/java/CheckstyleBearTest.py]: In that case you would do `from bears.java.CheckstyleBear import CheckstyleBear, known_checkstyle_or_path`

File: bears/js/JSONFormatBear.py (hunk 3 above)
Comment (AbdealiJK) [inline bears/js/JSONFormatBear.py]: It is already being tested. The coverage information is collated from all the python versions we test against - so, ther python 3.4 test will cover that branch

Comment (AbdealiJK) [inline bears/js/JSONFormatBear.py]: Ahh. Then you can't put the pytest --cov-fail-under condition :| because we do test for multiple branches. in fact we have some windows specific branches which are collated from the coverage on windows...

File: tests/java/CheckstyleBearTest.py [hunk 4]
```diff
@@ -1,18 +1,19 @@
 import os
 from queue import Queue
 
-from bears.java.CheckstyleBear import CheckstyleBear
+from bears.java import CheckstyleBear
 from tests.BearTestHelper import generate_skip_decorator
 from tests.LocalBearTestHelper import LocalBearTestHelper
 from coalib.settings.Section import Section
+from coalib.settings.Setting import path, Setting
```
Comment (underyx) [inline tests/java/CheckstyleBearTest.py]: stupid bear :-1:

File: .ci/deps.sh [hunk 5]
```diff
@@ -105,3 +105,9 @@ fi
 # Tailor (Swift) commands
 curl -fsSL https://tailor.sh/install.sh | sed 's/read -r CONTINUE < \/dev\/tty/CONTINUE=y/' > install.sh
 sudo bash install.sh
+
+# making coala cache the dependencies downloaded upon first run
+echo '' > dummy
+coala-ci --bears CheckstyleBear --files dummy --no-config --bear-dirs bears || true
+coala-ci --bears ScalaLintBear --files dummy --no-config --bear-dirs bears || true
+
```
Comment (sils1297) [inline .ci/deps.sh]: unneeded newline

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/cd92938f7977f06f30fe1273d2f2a9ecebc1b23a

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/89343f2824a24baa2005c188215ce5086cd10ec8

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/0fcc0e01e0f98f41ea33f896757c6717decf629a

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/03ed8567c2d20f5ced9111e7ca945d205f05f8ab

File: tests/java/CheckstyleBearTest.py [hunk 6]
```diff
@@ -1,18 +1,21 @@
...
 
 
-@generate_skip_decorator(CheckstyleBear)
+@generate_skip_decorator(CheckstyleBear.CheckstyleBear)
```
Comment (sils1297) [inline tests/java/CheckstyleBearTest.py]: can you just adjust the import?

Comment (underyx) [inline tests/java/CheckstyleBearTest.py]: How?

Comment (sils1297) [inline tests/java/CheckstyleBearTest.py]: `from ....CheckstyleBear import CheckstyleBear`

File: setup.cfg [hunk 7]
```diff
@@ -1,4 +1,4 @@
-[pytest]
+[tool:pytest]
```
Comment (sils1297) [inline setup.cfg]: unrelated change

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/e46b0ef17a0fd5c3998f1fed88b9179d6ca01bf9

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala-bears/pull/767/commits/c2871ecb3d6056f76d722ade8803d9156a785213

File: tests/js/JSComplexityBearTest.py [hunk 8]
```diff
@@ -1,4 +1,6 @@
-from bears.js.JSComplexityBear import JSComplexityBear
+import json
+
+from bears.js import JSComplexityBear
```
Comment (sils1297) [inline tests/js/JSComplexityBearTest.py]: why are you changing this import?

File: setup.cfg (hunk 7 above)
Comment (underyx) [inline setup.cfg]: I disagree, see commit message.

File: tests/java/CheckstyleBearTest.py (hunk 6 above)
Comment (underyx) [inline tests/java/CheckstyleBearTest.py]: But then it doesn't import the `known_checkstyle_or_path` function.

File: tests/js/JSComplexityBearTest.py (hunk 8 above)
Comment (underyx) [inline tests/js/JSComplexityBearTest.py]: To be able to access the `get_available_decodeerror` function.

File: bears/js/JSComplexityBear.py [hunk 9]
```diff
@@ -7,12 +7,21 @@
 from coalib.results.Result import Result
 
 
+def get_available_decodeerror():
+    try:
+        return json.decoder.JSONDecodeError
+    except AttributeError:  # Python 3.4 does not have JSONDecodeError there
+        return ValueError
```
Comment (sils1297) [inline bears/js/JSComplexityBear.py]: do we have an issue about the duplication with the other bear here?

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/31aaabecb7ceff07ef1a811644c5bad3d990355d https://github.com/coala-analyzer/coala-bears/pull/767/commits/dc0ccd9aa36f029766f7887bfaf67e62836af044 https://github.com/coala-analyzer/coala-bears/pull/767/commits/a76370d9cb03e4ad96d534bc5a3e933c44d781be https://github.com/coala-analyzer/coala-bears/pull/767/commits/ddc96427382e07de727549142401ce1f043a9d3a https://github.com/coala-analyzer/coala-bears/pull/767/commits/fec09fa0959663d26714c488f9e5ca4edc7c3a03 https://github.com/coala-analyzer/coala-bears/pull/767/commits/debc4d2ee1c5a1f7a3c166bc052b8770a4d77472 https://github.com/coala-analyzer/coala-bears/pull/767/commits/09d4a24d9a9a38ece26cc2b62ddbbc2a2bd20eab https://github.com/coala-analyzer/coala-bears/pull/767/commits/a273de0572860a110108400106a36f621d4e54a6 https://github.com/coala-analyzer/coala-bears/pull/767/commits/1523cb530d7df7807517ab034a948fccbb87f543 https://github.com/coala-analyzer/coala-bears/pull/767/commits/14ba8643232c2cf15569cd347fdd02c1b5625bd2

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/767/commits/ed65807c11e5f1803c2957ee81a5434ee778eb9b

**Comment by sils1297** [comment]
@rultor merge

**Comment by underyx** [comment]
@rultor please just try to merge it again :/

**Comment by underyx** [comment]
@rultor merge
